bonsai/UseCases/BudgetGateway.swift

```diff
@@ -0,0 +1,83 @@
+//
+//  BudgetGateway.swift
+//  bonsai
+//
+//  Created by hoang on 21.12.2021.
+//
+
+import Foundation
+import CoreData
+
+public protocol BudgetProtocol {
+   var budgetID: String { get }
+   var name: String { get }
+   var amount: Double { get }
+   var period: Int64 { get } // days ??
+}
+
+public final class BudgetGateway {
+
+   let mainContext: NSManagedObjectContext
+
+   init(mainContext: NSManagedObjectContext = DataController.sharedInstance.mainContext) {
+      self.mainContext = mainContext
+   }
+
+   @discardableResult
+   func create(name: String, amount: Double, period: Int64) throws -> Budget {
+      let budget = Budget(
+         context: mainContext,
+         name: name,
+         amount: amount,
+         period: period)
+      try mainContext.save()
+      return budget
+   }
+
+   func update(budget: Budget) throws {
+      try mainContext.save()
+   }
+
+   func delete(budget: Budget) throws {
+      mainContext.delete(budget)
+      try mainContext.save()
+   }
+
+   func getBudget(name: String) throws -> BudgetProtocol {
+      let request: NSFetchRequest<Budget> = Budget.fetchRequest()
+      request.fetchLimit = 1
+      request.predicate = NSPredicate(format: "name == %@", name)
```

Comment:
Predicates are tricky because unsafe. I have a few thoughts how we could improve it:
First suggestion is to use something like [this](https://gist.github.com/sisoje/f1444dff45618938ce81324a81316690) to add safety via swift KeyPaths
Another option is to use `NSPredicate.initWithBlock` as there we can use safe swift checks, however it might not work with SQL storage (needs to be checked if we use one)
Last option could be to use #keypaths like showed [there](https://nukedbit.dev/how-to-use-keypath-on-coredata-predicate-for-easy-refactoring-and-preventing-mistakes/)